concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+(We acknowledge that there's much more to "type theory" than this in the
+broader literature and history of computer science, but since IPLD is
+about *data*, focusing on this subset of type theory suits our purposes.)
+
+You'll find this document is linked to by the IPLD specs -- both for describing
+key theoretical foundations of the Data Model, and also to better describe some
+of the advanced features in the Schema layer (plus the reasons for them) --
+*and* you'll also find this document linked to by the
+[library design notes](../design/libraries) -- where we use some of this
+terminology to describe features of programming languages without resorting
+to language-specific vocabulary.
+
+
+Cardinality
+-----------
+
+"Cardinality" means "[the number of elements in a set](https://en.wikipedia.org/wiki/Cardinality)".
+
+In IPLD, we use the phrase "the cardinality of a type" to define how many members a type can have.
+
+### Why is Cardinality important?
+
+Cardinality is important to the way we define and think about our Data Model and Schema type systems,
+because if the cardinality of two parts of a model aren't the same cardinality,
+then that means one of them is less expressive.
+If we intended two systems to be interchangeable, but they have unequal cardinality,
+then we can quickly and clearly identify that there must be a bug somewhere!
+
+Cardinality concepts are also useful for clear documentation.
+
+Cardinality counting can be used as a design planning exercise:
+count how many states your types *can* have, and then consider how many states
+your code is designed to handle, and if they're not equal,
+then it's clear that you have work to do!
+
+In practical application, you can see some examples of cardinality-counting used
+to describe `nullable` vs `optional` vs `implicit` values in the IPLD Schemas documentation:
+[Cardinality Examples](../schemas/schema-kinds.md#cardinality-examples).
+These cardinality countings are both key to the documentation,
+and key to understanding and using those features to build unambiguous APIs.
+
+### How do we Count Cardinality?
+
+Counting cardinality is probably best done by example.
+
+Consider a boolean: it can have the value 'true', or it can have the value 'false'.
+We simply count these states up!  The cardinality of a boolean is `2`!
+
+Now, consider a pair of boolean values: how many states can it have?
+Four: `(true, true)`, `(true, false)`, `(false, true)`, or `(false, false)`.
+So, the cardinality of two booleans is `4`.
+
+The exercise gets more interesting as we consider larger accumulations of values.
+
+Consider three boolean values: how many states can it have?
+(You can draw the full table yourself, for this one.)
+When you count them all up, you should reach `8`.
+
+But now, let's imagine something new: suppose we've got some sort of values
+that we call "nooleans".  Nooleans can have the value 'nue' or 'fuu'.
```

Comment:
It seems like this example doesn't actually go anywhere? I was expecting you to explicitly note the isomorphism with bools.

Reply:
^ that. This needs to be expanded. Is it about "interchangeable systems" or something else? When I started reading this I was expecting this to be about booleans-with-null, introducing a cardinality of 3. Either way, this needs to have its dots connected.

Reply:
I don't honestly remember what was in my head when I started writing this bit.  I'm going to drop it entirely now in the interest of getting this document finished enough to merge gracefully.